scipy/spatial/transform/_rotation.pyx

```diff
@@ -94,6 +94,27 @@ cdef inline int _elementary_basis_index(uchar axis) noexcept:
     elif axis == b'y': return 1
     elif axis == b'z': return 2
 
+# Reduce the quaternion double coverage of the rotation group to a unique
+# canonical "positive" single cover
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef inline void _quat_canonical_single(double[:] q) nogil:
```

Comment:
Just FYI. This function will behave correctly when compiled in Cython3 but it will have slightly performance overhead:
<img width="1280" alt="image" src="https://github.com/scipy/scipy/assets/827060/134d2c45-f514-4b0c-8995-3a5e295f0f75">

This overhead will be even bigger when the function is called within `with nogil():` block (right now it is not this case)

Reply:
Here is Cython 0.29.X for comparison:
<img width="1318" alt="image" src="https://github.com/scipy/scipy/assets/827060/d6633a96-815a-4632-abd5-662433459fe4">